readthedocs/notifications/querysets.py

```diff
@@ -32,3 +36,53 @@ def add(self, *args, **kwargs):
             return notification
 
         return super().create(*args, attached_to=attached_to, **kwargs)
+
+    def for_user(self, user):
+        """
+        Retrieve all notifications related to a particular user.
+
+        Given a user, returns all the notifications that:
+
+         - are attached to an ``Organization`` where the user is owner
+         - are attached to a ``Project`` where the user is admin
+         - are attacehd to the ``User`` themselves
+        """
+        # Need to be here due to circular imports
+        from readthedocs.organizations.models import Organization
+        from readthedocs.projects.models import Project
+
+        # http://chibisov.github.io/drf-extensions/docs/#usage-with-generic-relations
+        user_notifications = self.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(User),
+            attached_to_id=user.pk,
+        )
+
+        # NOTE: using ``Subquery`` work better regarding performance since it makes only one query to the databse,
+        # instead of doing one to get the ids and another one to get the notifications.
+        # It also works better for users that have many projects/organizations since we don't need to load the ids
+        # in memory and then dump them into the SQL query.
+        # Besides, it performs better when applying the filters via ``NotificationFilter``
+        project_notifications = self.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Project),
+            attached_to_id__in=Subquery(
+                AdminPermission.projects(
+                    user,
+                    admin=True,
+                    member=False,
+                ).values("id")
+            ),
+        )
+        organization_notifications = self.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Organization),
+            attached_to_id__in=Subquery(
+                AdminPermission.organizations(
+                    user,
+                    admin=True,
+                    member=False,
+                ).values("id")
+            ),
+        )
```

Comment:
@stsewd You've done the most work here on query performance, any concerns on this pattern? For some background, this query will be executed on most/all dashboard page views while logged in.

Reply:
I think Django already detects this as a single query

https://github.com/readthedocs/readthedocs.org/pull/11009#discussion_r1457686522